### PR TITLE
Make expressions and constraints native values.

### PR DESCRIPTION
**Pull request by chriseth**
This is the first step in removing the "Custom" feature of the evaluator by making `expr` and `constr` native to the evaluator. It makes sense since those are also native types for the type checker.

### CODE REVIEW

File: book/src/pil/builtins.md [hunk 1]
```diff
@@ -99,6 +99,14 @@ if std::field::modulus() != 2**64 - 2**32 + 1 {
 };
 ```
 
+### Evaluate
+
+```rust
+let std::prover::eval: expr -> fe
+```
+
+TODO
```
Comment (chriseth) [inline book/src/pil/builtins.md:108]: Also mention this in` ## Declaring and Referencing Columns` in types.md

File: executor/src/witgen/query_processor.rs [hunk 2]
```diff
@@ -103,102 +96,33 @@ struct Symbols<'a, T: FieldElement> {
...
+            PolynomialType::Committed | PolynomialType::Intermediate => self
+                .rows
+                .get_value(&poly_ref)
+                .ok_or(EvalError::DataNotAvailable)?,
             PolynomialType::Constant => {
-                let values = self.fixed_data.fixed_cols[&function.poly_id].values;
-                values[(usize::try_from(row).unwrap() % values.len())
-                    .to_u64()
-                    .unwrap() as usize]
+                // TODO do we need this special case?
```
Comment (chriseth) [inline executor/src/witgen/query_processor.rs:121]: @georgwiese do you know if we can somehow access the fixed column values directyl from a row pair? Or do I need to go through fixed_data like here?

File: pil-analyzer/src/evaluator.rs [hunk 3]
```diff
@@ -655,31 +762,35 @@ mod internal {
         Ok(match b {
             BuiltinFunction::ArrayLen => match arguments.pop().unwrap().as_ref() {
                 Value::Array(arr) => Value::Integer((arr.len() as u64).into()),
-                v => Err(EvalError::TypeError(format!(
+                v => panic!(
```
Comment (georgwiese) [inline pil-analyzer/src/evaluator.rs:765]: Why this change?

File: pil-analyzer/src/evaluator.rs [hunk 4]
```diff
@@ -655,31 +762,35 @@ mod internal {
...
                     v.type_name()
-                )))?,
+                ),
             },
             BuiltinFunction::Panic => {
                 let msg = match arguments.pop().unwrap().as_ref() {
                     Value::String(msg) => msg.clone(),
-                    // As long as we do not yet have types, we just format any argument.
-                    x => x.to_string(),
+                    v => panic!(
```
Comment (georgwiese) [inline pil-analyzer/src/evaluator.rs:773]: Shouldn't this return an `EvalError::TypeError`?

File: pil-analyzer/src/evaluator.rs [hunk 5]
```diff
@@ -655,31 +762,35 @@ mod internal {
...
+                    ),
                 };
                 Err(EvalError::FailedAssertion(msg))?
             }
             BuiltinFunction::Print => {
                 let msg = match arguments.pop().unwrap().as_ref() {
                     Value::String(msg) => msg.clone(),
-                    // As long as we do not yet have types, we just format any argument.
-                    x => x.to_string(),
+                    v => panic!(
```
Comment (georgwiese) [inline pil-analyzer/src/evaluator.rs:783]: Same here

File: executor/src/witgen/query_processor.rs (hunk 2 above)
Comment (georgwiese) [reply executor/src/witgen/query_processor.rs:121]: As far as I can see, `RowPair::get_value` does not currently handle reference to fixed columns, but I think it should, as it already has access to `fixed_data`! `RowPair::evaluate` already handles this (but it's evaluating to an `AffineExpression`, which I guess we don't want to do here).

**Review by georgwiese — COMMENTED** [review]
Nice! I actually tried removing the `Custom` stuff and it works, it's not used anymore! But that can be a follow-up.

File: asm-to-pil/src/vm_to_constrained.rs [hunk 6]
```diff
@@ -952,11 +948,11 @@ impl<T: FieldElement> ASMPILConverter<T> {
                 let prover_query_arms = free_value_query_arms.remove(reg).unwrap();
                 let prover_query = (!prover_query_arms.is_empty()).then_some({
                     FunctionDefinition::Query(Expression::LambdaExpression(LambdaExpression {
-                        params: vec!["i".to_string()],
+                        params: vec!["__i".to_string()],
```
[the remark this answers is not in the document]
Comment (chriseth) [reply asm-to-pil/src/vm_to_constrained.rs:953]: True, but we can always use `std::prover::eval(line)` or something. But yeah, let's keep it for block machines, and not mention it for VMs, because there, we really don't want anything to depend on explicit row numbers.

File: executor/src/witgen/query_processor.rs (hunk 2 above)
Comment (chriseth) [reply executor/src/witgen/query_processor.rs:121]: I mean it depends. I'll remove the TODO here and keep it like that for now.

**Comment by chriseth** [comment]
> Nice! I actually tried removing the `Custom` stuff and it works, it's not used anymore! But that can be a follow-up.

Yes, it's in the queue: https://github.com/powdr-labs/powdr/pull/1116

File: pil-analyzer/src/evaluator.rs (hunk 3 above)
Comment (chriseth) [reply pil-analyzer/src/evaluator.rs:765]: Actually most of the type errors should be replaced by panic, since it's now an internal error in the type checker if this does not match up.

File: pil-analyzer/src/evaluator.rs (hunk 4 above)
Comment (chriseth) [reply pil-analyzer/src/evaluator.rs:773]: See above.

File: pil-analyzer/src/evaluator.rs (hunk 5 above)
Comment (chriseth) [reply pil-analyzer/src/evaluator.rs:783]: See above.

**Review by georgwiese — APPROVED** [review]
go go go
